### PR TITLE
[WIP] Fix/Suppress internal API usage in EFCore.Relational

### PR DESCRIPTION
**Pull request by smitpatel**
Major points I found
- IndentedStringBuilder
- ExpressionEqualityComparer
- MultiGraph
- Metadata printing
- DatabaseFacadeDependencies is seriously complicated
- ValueGeneratorSelector

### CODE REVIEW

File: src/EFCore.Relational/Extensions/RelationalDatabaseFacadeExtensions.cs [hunk 1]
```diff
@@ -203,8 +203,10 @@ public static IEnumerable<string> GetPendingMigrations([NotNull] this DatabaseFa
             Check.NotNull(parameters, nameof(parameters));
 
             var facadeDependencies = GetFacadeDependencies(databaseFacade);
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Extensions/RelationalDatabaseFacadeExtensions.cs]: facadeDependencies

File: src/EFCore.Relational/Extensions/RelationalDatabaseFacadeExtensions.cs [hunk 2]
```diff
@@ -218,7 +220,9 @@ public static IEnumerable<string> GetPendingMigrations([NotNull] this DatabaseFa
                             facadeDependencies.RelationalConnection,
                             rawSqlCommand.ParameterValues,
                             null,
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Extensions/RelationalDatabaseFacadeExtensions.cs]: IDatabaseFacadeDependenciesAccessor

File: src/EFCore.Relational/Extensions/RelationalPropertyExtensions.cs [hunk 3]
```diff
@@ -775,7 +775,9 @@ private static IProperty FindSharedObjectRootPrimaryKeyProperty([NotNull] IPrope
                     return null;
                 }
 
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Extensions/RelationalPropertyExtensions.cs]: IndexOf

File: src/EFCore.Relational/Infrastructure/EntityFrameworkRelationalServicesBuilder.cs [hunk 4]
```diff
@@ -54,6 +54,7 @@ public class EntityFrameworkRelationalServicesBuilder : EntityFrameworkServicesB
         public static readonly IDictionary<Type, ServiceCharacteristics> RelationalServices
             = new Dictionary<Type, ServiceCharacteristics>
             {
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Infrastructure/EntityFrameworkRelationalServicesBuilder.cs]: ServiceCharacteristics

File: src/EFCore.Relational/Infrastructure/EntityFrameworkRelationalServicesBuilder.cs [hunk 5]
```diff
@@ -114,7 +116,9 @@ protected override ServiceCharacteristics GetServiceCharacteristics(Type service
         {
             return RelationalServices.TryGetValue(serviceType, out var characteristics)
                 ? characteristics
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Infrastructure/EntityFrameworkRelationalServicesBuilder.cs]: GetServiceCharacteristics

File: src/EFCore.Relational/Infrastructure/EntityFrameworkRelationalServicesBuilder.cs [hunk 6]
```diff
@@ -171,6 +175,7 @@ public override EntityFrameworkServicesBuilder TryAddCoreServices()
             TryAdd<IRelationalQueryStringFactory, RelationalQueryStringFactory>();
 
             ServiceCollectionMap.GetInfrastructure()
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Infrastructure/EntityFrameworkRelationalServicesBuilder.cs]: AddDependencySingleton

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs [hunk 7]
```diff
@@ -348,8 +348,10 @@ private static bool IsIdentifyingPrincipal(IEntityType dependentEntityType, IEnt
                     {
                         var declaringType = tokenPair.Value.DeclaringEntityType;
                         if (!declaringType.IsAssignableFrom(entityType)
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: IsInOwnershipPath

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs [hunk 8]
```diff
@@ -373,7 +375,9 @@ private static bool IsIdentifyingPrincipal(IEntityType dependentEntityType, IEnt
                 {
                     foreach (var missingColumn in missingConcurrencyTokens)
                     {
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: GetAllBaseTypes

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs [hunk 9]
```diff
@@ -641,7 +645,9 @@ private static bool IsIdentifyingPrincipal(IEntityType dependentEntityType, IEnt
             foreach (var entityType in model.GetEntityTypes())
             {
                 if (entityType.BaseType != null
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: FindAnnotation

File: src/EFCore.Relational/Metadata/Conventions/SharedTableConvention.cs [hunk 10]
```diff
@@ -157,9 +157,13 @@ private static bool ShouldUniquify(IConventionEntityType entityType, ICollection
                     continue;
                 }
 
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Conventions/SharedTableConvention.cs]: GetIdentifyingMemberInfo

File: src/EFCore.Relational/Metadata/Conventions/TableNameFromDbSetConvention.cs [hunk 11]
```diff
@@ -26,7 +26,9 @@ public class TableNameFromDbSetConvention : IEntityTypeAddedConvention, IEntityT
             [NotNull] ProviderConventionSetBuilderDependencies dependencies,
             [NotNull] RelationalConventionSetBuilderDependencies relationalDependencies)
         {
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Conventions/TableNameFromDbSetConvention.cs]: CreateClrTypeDbSetMapping

File: src/EFCore.Relational/Metadata/Conventions/TableNameFromDbSetConvention.cs [hunk 12]
```diff
@@ -61,7 +63,9 @@ public class TableNameFromDbSetConvention : IEntityTypeAddedConvention, IEntityT
                 && entityType.ClrType != null
                 && _sets.ContainsKey(entityType.ClrType))
             {
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Conventions/TableNameFromDbSetConvention.cs]: Name

File: src/EFCore.Relational/Metadata/ITable.cs [hunk 13]
```diff
@@ -49,7 +49,9 @@ public interface ITable : ITableBase
         /// </summary>
         IEnumerable<ICheckConstraint> CheckConstraints
             => EntityTypeMappings.SelectMany(m => CheckConstraint.GetCheckConstraints(m.EntityType))
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/ITable.cs]: Distinct

File: src/EFCore.Relational/Metadata/Internal/CheckConstraint.cs [hunk 14]
```diff
@@ -146,7 +146,9 @@ public static IEnumerable<CheckConstraint> GetCheckConstraints([NotNull] IEntity
         /// </summary>
         public virtual void UpdateConfigurationSource(ConfigurationSource configurationSource)
         {
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/CheckConstraint.cs]: Max

File: src/EFCore.Relational/Metadata/Internal/Column.cs [hunk 15]
```diff
@@ -56,7 +56,9 @@ public Column([NotNull] string name, [CanBeNull] string type, [NotNull] Table ta
         ///     any release. You should only use it directly in your code with extreme caution and knowing that
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/Column.cs]: MetadataDebugStringOptions

File: src/EFCore.Relational/Metadata/Internal/ColumnExtensions.cs [hunk 16]
```diff
@@ -52,9 +54,13 @@ public static class ColumnExtensions
             builder.Append(")");
 
             if (!singleLine &&
+#pragma warning disable EF1001 // Internal EF Core API usage.
                 (options & MetadataDebugStringOptions.IncludeAnnotations) != 0)
+#pragma warning restore EF1001 // Internal EF Core API usage.
             {
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/ColumnExtensions.cs]: AnnotationsToDebugString

File: src/EFCore.Relational/Metadata/Internal/ForeignKeyConstraint.cs [hunk 17]
```diff
@@ -47,7 +47,9 @@ public class ForeignKeyConstraint : Annotatable, IForeignKeyConstraint
         ///     any release. You should only use it directly in your code with extreme caution and knowing that
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/ForeignKeyConstraint.cs]: ForeignKeyComparer

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs [hunk 18]
```diff
@@ -382,25 +382,33 @@ private static void PopulateInternalForeignKeys(TableBase table)
                 {
                     if (foreignKey.IsUnique
                         && foreignKey.PrincipalKey.IsPrimaryKey()
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/RelationalModel.cs]: IsIntraHierarchical

File: src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs [hunk 19]
```diff
@@ -275,8 +275,10 @@ public virtual IReadOnlyList<MigrationOperation> GetDifferences(IModel source, I
                 }
             }
 
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs]: Multigraph

File: src/EFCore.Relational/Migrations/Internal/Migrator.cs [hunk 20]
```diff
@@ -318,6 +318,7 @@ public virtual void Migrate(string targetMigration = null)
                 out var migrationsToRevert,
                 out var actualTargetMigration);
 
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Migrations/Internal/Migrator.cs]: IndentedStringBuilder

File: src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs [hunk 21]
```diff
@@ -99,7 +99,9 @@ protected virtual IUpdateSqlGenerator SqlGenerator
         /// <summary>
         ///     Gets a comparer that can be used to compare two product versions.
         /// </summary>
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs]: SemanticVersionComparer

File: src/EFCore.Relational/Query/FromSqlParameterApplyingExpressionVisitor.cs [hunk 22]
```diff
@@ -17,7 +17,9 @@ namespace Microsoft.EntityFrameworkCore.Query
     public class FromSqlParameterApplyingExpressionVisitor : ExpressionVisitor
     {
         private readonly IDictionary<FromSqlExpression, Expression> _visitedFromSqlExpressions
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/FromSqlParameterApplyingExpressionVisitor.cs]: ReferenceEqualityComparer

File: src/EFCore.Relational/Query/Internal/FromSqlQueryRootExpression.cs [hunk 23]
```diff
@@ -64,9 +64,13 @@ public override bool Equals(object obj)
         private bool Equals(FromSqlQueryRootExpression queryRootExpression)
             => base.Equals(queryRootExpression)
                 && string.Equals(Sql, queryRootExpression.Sql, StringComparison.OrdinalIgnoreCase)
+#pragma warning disable EF1001 // Internal EF Core API usage.
                 && ExpressionEqualityComparer.Instance.Equals(Argument, queryRootExpression.Argument);
+#pragma warning restore EF1001 // Internal EF Core API usage.
 
         public override int GetHashCode()
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/FromSqlQueryRootExpression.cs]: ExpressionEqualityComparer

File: src/EFCore.Relational/Query/RelationalEntityShaperExpression.cs [hunk 24]
```diff
@@ -63,7 +65,9 @@ protected override LambdaExpression GenerateMaterializationCondition(IEntityType
                             body = Condition(
                                 allNonPkProperties
                                     .Select(p => NotEqual(
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalEntityShaperExpression.cs]: GetIndex

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 25]
```diff
@@ -682,7 +682,9 @@ protected override ShapedQueryExpression TranslateOfType(ShapedQueryExpression s
                     return source;
                 }
 
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs]: GetAllBaseTypes

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 26]
```diff
@@ -1186,12 +1188,16 @@ private Expression TryExpand(Expression source, MemberIdentity member)
                     var innerSequenceType = innerShapedQuery.Type.TryGetSequenceType();
                     var correlationPredicateParameter = Expression.Parameter(innerSequenceType);
 
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs]: CreateKeyAccessExpression

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 27]
```diff
@@ -1306,7 +1316,9 @@ private Expression TryExpand(Expression source, MemberIdentity member)
                 IEntityType entityType, TableExpression table, bool nullable)
             {
                 var propertyExpressions = new Dictionary<IProperty, ColumnExpression>();
+#pragma warning disable EF1001 // Internal EF Core API usage.
                 foreach (var property in entityType.GetTypesInHierarchy().SelectMany(EntityTypeExtensions.GetDeclaredProperties))
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs]: GetTypesInHierarchy

File: src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs [hunk 28]
```diff
@@ -234,7 +234,9 @@ protected override Expression VisitExtension(Expression node)
                 if (node is SqlExpression sqlExpression
                     && !(node is SqlFragmentExpression)
                     && !(node is SqlFunctionExpression sqlFunctionExpression
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs]: IsQueryableType

File: src/EFCore.Relational/Storage/Internal/DbParameterCollectionExtensions.cs [hunk 29]
```diff
@@ -32,7 +32,9 @@ public static class DbParameterCollectionExtensions
             bool logParameterValues)
             => parameters
                 .Cast<DbParameter>()
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Storage/Internal/DbParameterCollectionExtensions.cs]: Join

File: src/EFCore.Relational/Storage/RelationalCommandBuilder.cs [hunk 30]
```diff
@@ -20,7 +20,9 @@ namespace Microsoft.EntityFrameworkCore.Storage
     public class RelationalCommandBuilder : IRelationalCommandBuilder
     {
         private readonly List<IRelationalParameter> _parameters = new List<IRelationalParameter>();
+#pragma warning disable EF1001 // Internal EF Core API usage.
```
Comment (smitpatel) [inline src/EFCore.Relational/Storage/RelationalCommandBuilder.cs]: IndentedStringBuilder

File: src/EFCore.Relational/Extensions/RelationalPropertyExtensions.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Extensions/RelationalPropertyExtensions.cs]: Can we make all `EnumerableExtensions` shared internal?

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs (hunk 7 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: Can be public, needs better name

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs (hunk 8 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: Can be public, needs better name

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs (hunk 9 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: Cast to IConventionEntityType

File: src/EFCore.Relational/Metadata/Conventions/SharedTableConvention.cs (hunk 10 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Conventions/SharedTableConvention.cs]: Can inline

File: src/EFCore.Relational/Metadata/Internal/Column.cs (hunk 15 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Internal/Column.cs]: We should probably move it to .Infrastructure

File: src/EFCore.Relational/Metadata/Internal/ColumnExtensions.cs (hunk 16 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Internal/ColumnExtensions.cs]: Can expose through a public extension

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs (hunk 18 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Internal/RelationalModel.cs]: Can be public, needs better name

File: src/EFCore.Relational/Metadata/Internal/CheckConstraint.cs (hunk 14 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Internal/CheckConstraint.cs]: Can be public

File: src/EFCore.Relational/Metadata/Internal/Column.cs (hunk 15 above)
Comment (smitpatel) [reply src/EFCore.Relational/Metadata/Internal/Column.cs]: Filed https://github.com/dotnet/efcore/issues/20381

**Pull request closed without merging**